src/main/java/net/dries007/tfc/objects/container/ContainerKnapping.java

```diff
@@ -130,6 +130,30 @@ private void resetMatrix()
         consumeIngredientStackAfterComplete();
     }
 
+    /**
+     * Used in client to check a slot state in the matrix
+     * JEI won't cause issues anymore see https://github.com/TerraFirmaCraft/TerraFirmaCraft/issues/718
+     *
+     * @param index the slot index
+     * @return the boolean state for the checked slot
+     */
+    public boolean getSlotState(int index)
+    {
+        return matrix.get(index);
+    }
+
+    /**
+     * Used in client to set a slot state in the matrix
+     * JEI won't cause issues anymore see https://github.com/TerraFirmaCraft/TerraFirmaCraft/issues/718
+     *
+     * @param index the slot index
+     * @param value the value you wish to set the state to
+     */
+    public void setSlotState(int index, boolean value)
+    {
+        matrix.set(index, value);
+    }
```

Comment:
Nowhere do these sync the server matrix state to the client, is that intentional? i.e. the client is tracking it's matrix completely independently of the server.

Reply:
yes, that's intentional. We don't need to sync the server matrix->client because that would be redundant/cause lag issues. While recipe logic is done on server, the client is the one who draws the matrix (sending packets to server on which slot he clicked).

The lag issue i mentioned would be like block breaking lag: Client click on slot, slot is sent to server, server send to client a resync packet while client already clicked at another slot, the slot is visibly returning to true state before going false again.